src/pnpq/devices/waveplate_stub.py

```diff
@@ -0,0 +1,219 @@
+        self.__ensure_port_open()
+        self.logger.info("Stub Waveplate Disable Channel: %s", chanid)
+
+        if chanid > self.max_channel:
+            raise WaveplateInvalidMotorChannelError(f"Invalid motor channel: {chanid}.")
+
+        # Remove channel from enabled channels
+        self.enabled_channels.remove(chanid)
+
+        # Simulated delay
+        time.sleep(0.1)
+
+    def enable_channel(self, chanid: int) -> None:
+        """Enable a channel on the device specified by chanid"""
+        self.__ensure_port_open()
+        self.logger.info("Stub Waveplate Enable Channel: %s", chanid)
+        if chanid > self.max_channel:
+            raise WaveplateInvalidMotorChannelError(f"Invalid motor channel: {chanid}.")
+
+        self.enabled_channels.add(chanid)
+
+        # Simulated delay
+        time.sleep(0.1)
+
+        # TODO: Return a fake reply from the device
+
+    def device_resolution(self) -> int:
+        """Get the device resolution"""
+        return self.resolution
+
+    def getpos(self) -> int:
+        """Get the current position of the device in steps"""
+        self.__ensure_port_open()
+        self.logger.info("Stub Waveplate Get Position")
+        self.logger.info("Current Position: Steps: %s Degrees: %s", self.current_position, self.current_position / self.resolution)
+        return self.current_position
+
+    def get_degree(self) -> float:
+        """Get the current position of the device in degrees"""
+        self.__ensure_port_open()
+        self.logger.info("Stub Waveplate Get Degree")
+        return self.current_position / self.resolution
+
+    def rotate(self, degree: int | float) -> None:
+        """Rotate the device to a specified absolute degree"""
+        self.__ensure_port_open()
+        self.__ensure_valid_degree(degree)
+
+        if not self.__stub_check_channel(1):
+            # Do nothing if channel is not enabled
+            return
+
+        self.logger.info("Stub Waveplate Rotate to %s", degree)
+        # Calculate number of steps to move (round to nearest integer)
+        move_position = int(degree * self.resolution)
+        # Update current position
+        self.__set_steps(move_position)
+        # Delay to simulate rotation (for now: v=1ms/deg)
+        time.sleep(abs(move_position - self.current_position) / 1000)
+        # TODO: Return a fake reply from the device
+
+    def step_backward(self, steps: int) -> None:
+        """Step backward by a specified number of steps"""
+        # Get new position and set steps
+        new_steps = self.current_position - steps
+        self.__set_steps(new_steps)
+
+    def step_forward(self, steps: int) -> None:
+        """Step forward by a specified number of steps"""
+        # Get new position and set steps
+        new_steps = self.current_position + steps
+        self.__set_steps(new_steps)
+
+    def rotate_relative(self, degree: int | float) -> None:
```

Comment:
So, as far as I know, type hints are [only used by static analysis tools, not the interpreter](https://docs.python.org/3/library/typing.html). If you are using VSCode and it's complaining about this line when only `float` is set instead of `int | float`, I would be curious to see the error.

I have realized that I need to do more research on using static analysis tools before I understand which is the "right" way to do this in Python. My intuition might be wrong.

Reply:
The actual driver had this too (albeit for `rotate`, there's no type hint for `rotate_relative`), we may need to refactor this.